# Group 16
### Group Member details:
*   Ajay Kumar (2023AC05990) - 100% Contribution
*   Rajveer Mathur (2023AC05076) - 100% Contribution
*   Jasman Arora (2023AC05300) - 100% Contribution

### Step 1: Environment Setup

In [1]:
# Libaries required for the code
import os
import pandas as pd
import librosa
import numpy as np
import soundfile as sf
from sklearn.model_selection import train_test_split
from sklearn.preprocessing import StandardScaler
import tabulate
from sklearn.svm import SVC
from sklearn.metrics import accuracy_score, classification_report, confusion_matrix

# Remove the existing directory if it exists
import shutil
if os.path.exists('Wake-Word-Classifier-ASR'):
    shutil.rmtree('Wake-Word-Classifier-ASR')

# Clone the GitHub repository to access the data locally
!git clone https://github.com/Rajveermathur/Wake-Word-Classifier-ASR.git

base_dir = "Wake-Word-Classifier-ASR/data"
alexa_dir = 'Wake-Word-Classifier-ASR/data/alexa'
jarvis_dir = 'Wake-Word-Classifier-ASR/data/jarvis'

Cloning into 'Wake-Word-Classifier-ASR'...
remote: Enumerating objects: 114, done.
remote: Counting objects: 100% (114/114), done.
remote: Compressing objects: 100% (109/109), done.
remote: Total 114 (delta 30), reused 73 (delta 4), pack-reused 0 (from 0)
Receiving objects: 100% (114/114), 2.14 MiB | 11.80 MiB/s, done.
Resolving deltas: 100% (30/30), done.


In [2]:
# Check if the directories exist
# Alexa dataset directory, files check and verification
for filename in os.listdir(alexa_dir):
    if filename.endswith('.flac'):
        file_path = os.path.join(alexa_dir, filename)
        try:
            y, sr = librosa.load(file_path, sr=None)  # sr=None keeps the original sample rate
            print(f"✅ Loaded: {filename} | Sample Rate: {sr} | Duration: {len(y)/sr:.2f}s")
        except Exception as e:
            print(f"❌ Could not load {filename}: {e}")

✅ Loaded: 1.flac | Sample Rate: 16000 | Duration: 3.66s
✅ Loaded: 26.flac | Sample Rate: 16000 | Duration: 2.74s
✅ Loaded: 24.flac | Sample Rate: 16000 | Duration: 2.10s
✅ Loaded: 25.flac | Sample Rate: 16000 | Duration: 2.98s
✅ Loaded: 16.flac | Sample Rate: 16000 | Duration: 1.94s
✅ Loaded: 5.flac | Sample Rate: 16000 | Duration: 3.18s
✅ Loaded: 11.flac | Sample Rate: 16000 | Duration: 2.02s
✅ Loaded: 2.flac | Sample Rate: 16000 | Duration: 2.42s
✅ Loaded: 18.flac | Sample Rate: 16000 | Duration: 3.78s
✅ Loaded: 7.flac | Sample Rate: 16000 | Duration: 3.70s
✅ Loaded: 20.flac | Sample Rate: 16000 | Duration: 1.74s
✅ Loaded: 17.flac | Sample Rate: 16000 | Duration: 2.38s
✅ Loaded: 9.flac | Sample Rate: 16000 | Duration: 2.30s
✅ Loaded: 10.flac | Sample Rate: 16000 | Duration: 2.02s
✅ Loaded: 14.flac | Sample Rate: 16000 | Duration: 2.22s
✅ Loaded: 21.flac | Sample Rate: 16000 | Duration: 2.22s
✅ Loaded: 19.flac | Sample Rate: 16000 | Duration: 1.70s
✅ Loaded: 22.flac | Sample Rate: 160

In [3]:
# Check if the directories exist
# Jarvis dataset directory, files check and verification
for filename in os.listdir(jarvis_dir):
    if filename.endswith('.wav'):
        file_path = os.path.join(jarvis_dir, filename)
        try:
            y, sr = librosa.load(file_path, sr=None)  # sr=None preserves original sampling rate
            print(f"✅ Loaded: {filename} | Sample Rate: {sr} | Duration: {len(y)/sr:.2f}s")
        except Exception as e:
            print(f"❌ Could not load {filename}: {e}")


✅ Loaded: 1e128153-2111-4a6f-8792-6c24220ee5b0.wav | Sample Rate: 16000 | Duration: 2.53s
✅ Loaded: 1ddadbc1-4d52-4df8-88c2-8615587e2608.wav | Sample Rate: 16000 | Duration: 3.07s
✅ Loaded: 0df7b7ff-d667-488d-b4c2-ea78b4b847fe.wav | Sample Rate: 16000 | Duration: 3.07s
✅ Loaded: 00aba123-ae3a-4e0a-8603-9f7277b7d41f.wav | Sample Rate: 16000 | Duration: 3.07s
✅ Loaded: 0c7910b3-4414-4d48-9e6f-292a1d099f6f.wav | Sample Rate: 16000 | Duration: 3.07s
✅ Loaded: 1c9fda58-050b-4b8f-8289-dd08c3107c8c.wav | Sample Rate: 16000 | Duration: 3.07s
✅ Loaded: 1c9eaae2-1f5f-48bc-8379-6d1e1943f50b.wav | Sample Rate: 16000 | Duration: 3.07s
✅ Loaded: 1f549862-de22-4af0-875d-7ff927d127c2.wav | Sample Rate: 16000 | Duration: 3.07s
✅ Loaded: 00a97647-55b9-4f62-be20-8e4b0ee510b0.wav | Sample Rate: 16000 | Duration: 3.07s
✅ Loaded: 2a713afc-d208-4eab-9dc1-fcdd862d8db6.wav | Sample Rate: 16000 | Duration: 3.07s
✅ Loaded: 0e165e17-134f-4cee-9ec6-b43d1412d7d7.wav | Sample Rate: 16000 | Duration: 3.07s
✅ Loaded: 

### Step 2: Dataset Creation

In [4]:
# Create a DataFrame to hold the file paths and labels
data = []

# Iterate through the directories and collect file paths and labels
for wake_word in os.listdir(base_dir):
    folder_path = os.path.join(base_dir, wake_word)
    if os.path.isdir(folder_path):
        for filename in os.listdir(folder_path):
            if (wake_word == 'alexa' and filename.endswith('.flac')) or \
               (wake_word == 'jarvis' and filename.endswith('.wav')):
                file_path = os.path.join(folder_path, filename)
                data.append({'filepath': file_path, 'wake_word': wake_word})

# print(data)
df = pd.DataFrame(data)
print(df.head())

# Define 2 labels
df['wake_word'].value_counts()

# Convert wake words to binary labels
df['label'] = df['wake_word'].apply(lambda x: 0 if x == 'alexa' else 1)
print(df['label'].value_counts())
df.head()


                                      filepath wake_word
0   Wake-Word-Classifier-ASR/data/alexa/1.flac     alexa
1  Wake-Word-Classifier-ASR/data/alexa/26.flac     alexa
2  Wake-Word-Classifier-ASR/data/alexa/24.flac     alexa
3  Wake-Word-Classifier-ASR/data/alexa/25.flac     alexa
4  Wake-Word-Classifier-ASR/data/alexa/16.flac     alexa
label
0    30
1    30
Name: count, dtype: int64


,filepath,wake_word,label
0,Wake-Word-Classifier-ASR/data/alexa/1.flac,alexa,0
1,Wake-Word-Classifier-ASR/data/alexa/26.flac,alexa,0
2,Wake-Word-Classifier-ASR/data/alexa/24.flac,alexa,0
3,Wake-Word-Classifier-ASR/data/alexa/25.flac,alexa,0
4,Wake-Word-Classifier-ASR/data/alexa/16.flac,alexa,0


### Step 3: Feature Extraction


In [5]:
# Parameters
max_duration = 4.0
sample_rate = 16000
n_mfcc = 13

# Function to extract MFCC features
def extract_features(path):
    # Load audio file with librosa
    y, sr = librosa.load(path, sr=sample_rate, duration=max_duration)
    # Ensure the audio is of the expected length
    expected_len = int(max_duration * sample_rate)
    y = librosa.util.fix_length(y, size=expected_len)  # ← Fixed this line
    # Extract MFCC features
    mfcc = librosa.feature.mfcc(y=y, sr=sr, n_mfcc=n_mfcc)
    return np.mean(mfcc, axis=1)

# Extract features for each file
features = []
valid_labels = []

# Iterate through the DataFrame and extract features
for i, row in df.iterrows():
    try:
        mfcc_feat = extract_features(row['filepath'])
        features.append(mfcc_feat)
        valid_labels.append(row['label'])
    except Exception as e:
        print(f"❌ Error loading {row['filepath']}: {e}")

# Convert to arrays
X = np.array(features)
y = np.array(valid_labels)

# Normalize
scaler = StandardScaler()
X = scaler.fit_transform(X)

print('Feature Count:',len(X),'\n\nFirst 5 features:\n\n',X[0:5])

# Split into train and test sets (70/30)
X_train, X_test, y_train, y_test = train_test_split(
    X, y, test_size=0.3, random_state=42, stratify=y
)



Feature Count: 60 

First 5 features:

 [[-8.2311791e-01  2.3561766e+00  1.4166753e+00  7.9719609e-01
   8.5369986e-01  2.5227540e-03 -3.2840675e-01 -1.5845518e+00
  -1.0933896e+00 -1.5841948e+00 -2.4135475e+00 -2.7376866e+00
  -2.5435724e+00]
 [-1.2047522e+00 -9.7238868e-02 -4.1785832e-02  3.0646712e-01
   5.3339362e-01  8.7389088e-01  3.3323428e-01  8.1335920e-01
   8.6395383e-01  2.6705781e-01 -4.0392365e-02  2.1940045e-01
   1.7277631e-01]
 [-8.3026671e-01 -3.0684325e-01 -8.1312901e-01 -8.7117575e-02
  -3.8728174e-02  5.8238089e-01 -2.3227168e-02  5.7612723e-01
   8.5208970e-01  2.2018169e-01  1.5886332e-01  2.5281629e-01
   8.0715775e-01]
 [-4.5197228e-01 -4.6739662e-03 -3.1439462e-01  8.9722769e-03
  -9.9445231e-02  7.6376426e-01  2.4416888e-01  1.0066992e+00
   9.6103221e-01  4.2665797e-01  1.0272753e-01  2.2693190e-01
   6.8044758e-01]
 [ 2.0038350e+00 -1.7563424e-01 -8.9126307e-01 -2.0075672e+00
  -1.1769757e+00 -1.1533371e+00 -3.2673541e-01 -2.0334559e-02
   3.1253975e-02  5.

### Step 4: Train Wake Word Classifier

In [6]:
# Train a Support Vector Machine (SVM) classifier
clf = SVC(kernel='linear', C=1.0, random_state=42)
# Fit the model on the training data
clf.fit(X_train, y_train)

# Predict on test set
y_pred = clf.predict(X_test)

### Step 5: Evaluate Model

In [7]:
# Confusion matrix
cm = confusion_matrix(y_test, y_pred)

# Accuracy
accuracy = accuracy_score(y_test, y_pred) * 100

# Create table for tabulate
headers = ["Predicted \\ Actual", "Alexa", "Jarvis"]
table = [
    ["Alexa", cm[0][0], cm[0][1]],
    ["Jarvis", cm[1][0], cm[1][1]]
]

# Display
print("📊 Classification Report:\n", classification_report(y_test, y_pred, target_names=['alexa', 'jarvis']))
print('Confusion Matrix:\n',tabulate.tabulate(table, headers=headers, tablefmt="grid"))
print(f"\n✅ Overall Accuracy: {accuracy:.2f}%")

📊 Classification Report:
               precision    recall  f1-score   support

       alexa       1.00      0.78      0.88         9
      jarvis       0.82      1.00      0.90         9

    accuracy                           0.89        18
   macro avg       0.91      0.89      0.89        18
weighted avg       0.91      0.89      0.89        18

Confusion Matrix:
 +----------------------+---------+----------+
| Predicted \ Actual   |   Alexa |   Jarvis |
+======================+=========+==========+
| Alexa                |       7 |        2 |
+----------------------+---------+----------+
| Jarvis               |       0 |        9 |
+----------------------+---------+----------+

✅ Overall Accuracy: 88.89%


#### Short *Summary*



*   Which wake word is easier to classify?

Jarvis is easier to classify.
Recall = 1.00 → all Jarvis samples were correctly identified.
Precision = 0.82 → a few Alexa samples were misclassified as Jarvis.
For Alexa, recall is only 0.78, meaning 22% of Alexa samples were missed.
So, the model is more reliable in detecting Jarvis than Alexa.

*   Any misclassification patterns?

From the confusion matrix:
Alexa (9 actual samples): 7 were correctly classified as Alexa, but 2 were misclassified as Jarvis.
Jarvis (9 actual samples): all 9 were correctly classified (0 errors).

Pattern: Misclassification only happens when Alexa → Jarvis, never the other way.
This suggests that Alexa's MFCC features occasionally overlap with Jarvis's, making Alexa harder to distinguish.
